### PR TITLE
Support RemoveUnused in Scala 3.4+

### PR DESCRIPTION
**Pull request by bjaglin**
First half of https://github.com/scalacenter/scalafix/issues/1682

https://github.com/lampepfl/dotty/pull/16157

### CODE REVIEW

**Comment by tgodzik** [comment]
> Trying to understand why unused warnings don't show up as diagnostics in the SemanticDB files produced by dotty
> 
> ```
> ➜  cat Test.scala                                                                       
> //> using scala 3.3.0-RC6
> //> using options -Wunused:all -Ysemanticdb
> 
> import java.util.List
> 
> object Test
> ```
> 
> ```
> ➜  scala-cli Test.scala       
> [warn] ./Test.scala:4:18
> [warn] unused import
> [warn] import java.util.List
> [warn]                  ^^^^
> [error]  No main class found
> ```
> 
> ```
> ➜  metap Test.scala.semanticdb
> tmp/Test.scala
> --------------
> 
> Summary:
> Schema => SemanticDB v4
> Uri => tmp/Test.scala
> Text => empty
> Language => Scala
> Symbols => 1 entries
> Occurrences => 4 entries
> 
> Symbols:
> _empty_/Test. => final object Test extends Object { self: Test.type => +1 decls }
> 
> Occurrences:
> [3:7..3:11) => java/
> [3:12..3:16) => java/util/
> [3:17..3:21) => java/util/List#
> [5:7..5:11) <= _empty_/Test.
> ```

Looks like we don't actually write them :| I will ask someone to take a look.

**Comment by bjaglin** [comment]
> Looks like we don't actually write them :| I will ask someone to take a look.

@tgodzik I came to the same (sad) conclusion and was about to open a ticket on dotty

**Comment by bjaglin** [comment]
https://github.com/lampepfl/dotty/issues/17535

**Comment by anasbarg** [comment]
Any updates regarding this PR?

**Comment by tgodzik** [comment]
I think we are waiting for the new Scala 3 release

**Comment by bjaglin** [comment]
> I think we are waiting for the new Scala 3 release

Indeed, as soon as I see activity on the release train, I'll put that back in shape (and align with https://github.com/scalacenter/scalafix/pull/1800). The dotty PR [has been marked as `backport:nominated`](https://github.com/lampepfl/dotty/pull/17835#event-10714462429), so hopefully this makes it to Scala 3.3 LTS :crossed_fingers:

**Comment by bjaglin** [comment]
From the [dotty PR](https://github.com/lampepfl/dotty/pull/17835#issuecomment-1862460194)

> This particular change will land in 3.3.3 (that should have parity with 3.4.0). It will be available as an RC in late January and as a stable release in early March.

File: docs/rules/RemoveUnused.md [hunk 1]
```diff
@@ -67,7 +64,7 @@ object Main {
 }
 ```
 
-Remove unused pattern match variables:
+Remove unused pattern match variables (requires `-Wunused:unsafe-warn-patvars` on Scala 3):
```
Comment (bjaglin) [inline docs/rules/RemoveUnused.md:67]: https://github.com/lampepfl/dotty/pull/16157#issuecomment-1371943255

File: docs/users/installation.md [hunk 2]
```diff
@@ -9,7 +9,7 @@ title: Installation
 
 **Java LTS (8, 11, 17 or 21)**
 
-**Scala 2.12, 2.13 or 3.x** (all rules are not available for Scala 3.x)
```
Comment (bjaglin) [inline docs/users/installation.md:12]: only `ExplicitResultTypes` remains unsupported, and it's documented in rule docs, so I think we can drop this verbose disclaimer

File: scalafix-rules/src/main/scala/scalafix/internal/rule/RemoveUnused.scala [hunk 3]
```diff
@@ -26,26 +26,31 @@ class RemoveUnused(config: RemoveUnusedConfig)
...
   override def isRewrite: Boolean = true
 
   private def warnUnusedPrefix = List("-Wunused", "-Ywarn-unused")
   private def warnUnusedString = List("-Xlint", "-Xlint:unused")
   override def withConfiguration(config: Configuration): Configured[Rule] = {
+    def diagnosticsUnavailableInSemanticdb(version: String) =
+      version.startsWith("3.0") ||
+        version.startsWith("3.1") ||
+        version.startsWith("3.2") ||
+        version.startsWith("3.3")
```
Comment (bjaglin) [inline scalafix-rules/src/main/scala/scalafix/internal/rule/RemoveUnused.scala:39]: [3.3.3 is likely to get the semanticdb fix](https://github.com/lampepfl/dotty/pull/17835#issuecomment-1862460194) but I'll update when I see it happening in a RC

File: scalafix-rules/src/main/scala/scalafix/internal/rule/RemoveUnused.scala [hunk 4]
```diff
@@ -116,29 +128,33 @@ class RemoveUnused(config: RemoveUnusedConfig)
           .atomic
       }
 
+      def isUnusedImportee(importee: Importee): Boolean = {
+        val pos = importee match {
+          case Importee.Rename(from, _) => from.pos
+          case _ => importee.pos
+        }
+        isUnusedImport.exists { unused =>
+          unused.start <= pos.start && pos.end <= unused.end
```
Comment (bjaglin) [inline scalafix-rules/src/main/scala/scalafix/internal/rule/RemoveUnused.scala:137]: scala 3 diagnostics can be wider than scala 2 ones

File: docs/rules/RemoveUnused.md [hunk 5]
```diff
@@ -87,7 +84,7 @@ object Main {
 }
 ```
 
-Remove unused function parameters:
+Remove unused function parameters (Scala 2 only):
```
Comment (bjaglin) [inline docs/rules/RemoveUnused.md:87]: Despite the presence of `-Wunused:params` and the explicit mention of it in https://github.com/lampepfl/dotty/pull/16157, I did not manage to get diagnostics (nor warnings), so I am leaving this out for now

```
$ metap ./scalafix-tests/input/target/jvm-3.4.0-RC2/meta/META-INF/semanticdb/scalafix-tests/input/src/main/scala/test/removeUnused/RemoveUnusedParams.scala.semanticdb 
scalafix-tests/input/src/main/scala/test/removeUnused/RemoveUnusedParams.scala
------------------------------------------------------------------------------

Summary:
Schema => SemanticDB v4
Uri => scalafix-tests/input/src/main/scala/test/removeUnused/RemoveUnusedParams.scala
Text => empty
Language => Scala
Symbols => 11 entries
Occurrences => 28 entries

Symbols:
local0 => param unused: String
local1 => param unused: String
local2 => param used: String
local3 => param unused: Long
local4 => implicit param string: String
test/removeUnused/UnusedParams. => final object UnusedParams extends Object { self: UnusedParams.type => +5 decls }
test/removeUnused/UnusedParams.f. => val method fFunction1[String, Unit]
test/removeUnused/UnusedParams.ff. => val method ffFunction1[String, Unit]
test/removeUnused/UnusedParams.fs. => val method fsFunction2[String, Long, Unit]
test/removeUnused/UnusedParams.g(). => method g(x: Function1[String, Unit]): Unit
test/removeUnused/UnusedParams.g().(x) => param x: Function1[String, Unit]

Occurrences:
[3:8..3:12) => test/
[3:13..3:25) <= test/removeUnused/
[5:7..5:19) <= test/removeUnused/UnusedParams.
[6:6..6:7) <= test/removeUnused/UnusedParams.f.
[6:9..6:15) => scala/Predef.String#
[6:19..6:23) => scala/Unit#
[6:26..6:32) <= local0
[6:36..6:43) => scala/Predef.println(+1).
[7:6..7:8) <= test/removeUnused/UnusedParams.ff.
[7:12..7:18) <= local1
[7:20..7:26) => scala/Predef.String#
[7:31..7:38) => scala/Predef.println(+1).
[8:6..8:8) <= test/removeUnused/UnusedParams.fs.
[8:12..8:16) <= local2
[8:18..8:24) => scala/Predef.String#
[8:26..8:32) <= local3
[8:34..8:38) => scala/Long#
[8:43..8:50) => scala/Predef.println(+1).
[8:51..8:55) => local2
[9:6..9:7) <= test/removeUnused/UnusedParams.g().
[9:8..9:9) <= test/removeUnused/UnusedParams.g().(x)
[9:11..9:17) => scala/Predef.String#
[9:21..9:25) => scala/Unit#
[9:28..9:32) => scala/Unit#
[9:35..9:38) => scala/Predef.`???`().
[10:2..10:3) => test/removeUnused/UnusedParams.g().
[10:13..10:19) <= local4
[10:23..10:30) => scala/Predef.println(+1).

```

File: docs/developers/symbol-information.md [hunk 6]
```diff
@@ -107,6 +107,7 @@ Use `MethodSignature.parameterLists` to look up parameters of a method.
 def printMethodParameters(symbol: Symbol): Unit = {
   symbol.info.get.signature match {
     case signature @ MethodSignature(typeParameters, parameterLists, _) =>
+      println("signature  = " + signature)
```
Comment (bjaglin) [inline docs/developers/symbol-information.md:110]: [changes in `ScalafixBuild`](https://github.com/scalacenter/scalafix/pull/1728/files#diff-cf70a614681abb9e33115a49e0b4a52865bdfcdb24c4c0df4af6546d92f88f5f) enabled more unused warnings on the builds

File: docs/rules/RemoveUnused.md (hunk 5 above)
Comment (bjaglin) [reply docs/rules/RemoveUnused.md:87]: https://github.com/scalacenter/scalafix/issues/1937

**Pull request merged**